sycl/doc/extensions/AsyncGroupCopy/async_group_copy.asciidoc

```diff
@@ -0,0 +1,160 @@
+below. Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro’s value
+to determine which of the extension’s APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension implementation.
+|===
+
+
+## New `async_group_copy` function.
+`async_group_copy` will work very similarly to `async_work_group_copy`, but
+with the ability to utilize individual sub-groups. The group will be provided
+as a template parameter, in the same fashion as `group_barrier`.
+`async_group_copy` is intended as a potential optimization, and may add more
+overhead than a simple synchronous implementation in some cases.
+
+`async_group_copy` will asynchronously copy a given number of values of type
+`dataT` from global memory to local memory or from local memory to global
+memory. If an asynchronous copy is not supported then `async_group_copy` will
+fall back on a synchronous copy. `async_group_copy` is a free function with the
+following properties:
+
+* The work-items will not necessarily be synchronized, so must have a
+  consistent view of the memory before the invocation.
+* The order of operations is not specified.
+* All work-items of the group are required to call the function with the same
+  arguments, otherwise the behaviour is undefined.
+* All work-items of the group are required to call the function in convergent
+  control flow, otherwise the behaviour is undefined.
+* The type of `dataT` should be a scalar or vector type.
+* The data written by `async_group_copy` should not be read until `wait_for`
+  has been called on the returned `device_event`.
+
+The structs `dest_stride` and `src_stride` have been introduced to reduce errors
+when specifying the stride.
+
+NOTE: The order of the destination and source arguments differs from the SYCL
+1.2.1 `async_work_group_copy` function to be more in line with other copy
+functions in SYCL 2020.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+struct dest_stride {
+       std::size_t value;
+};
+
+struct src_stride {
+       std::size_t value;
+};
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> src, sycl::decorated_global_ptr<dataT> dest, size_t count);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> src, sycl::decorated_local_ptr<dataT> dest, size_t count);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> src, sycl::decorated_global_ptr<dataT> dest, size_t count, dest_stride destStride);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> src, sycl::decorated_local_ptr<dataT> dest, size_t count, src_stride srcStride);
+} // namespace sycl::ext::oneapi::experimental
+```
```

Comment:
I think there should be some description here about the operation that is performed by these functions.  The text above lists the restrictions, but it doesn't really say what they do.  In particular, I think the following should be clarified:

* That all work items calling the function collectively copy the elements in the range.
* A description of how the `destStride` and `srcStride` work.

Also, I assume that the number of work items in the group need not be the same as the number of elements that are copied, right?  Assuming this is the case, it would be good to clarify that too.

For example, maybe something like:

> When work items call one of these functions in converged control flow, all work items collectively copy `count` data elements from `src` to `dest`.  The number of work items in the group need not be the same as `count`.  When `destStride` is provided, contiguous elements from `src` are copied to potentially discontiguous locations in `dest`.  The element at `src[0]` is copied to the location at `dest[0*desStride]`, the element at `src[1]` is copied to the location at `dest[1*destStride]`, etc.  When `srcStride` is provided, elements that are potentially discontiguous in `src` are copied to contiguous locations in `dest` using a similar scheme.

Also, can `destStride` or `srcStride` be `0`?  If this is forbidden, it should be listed as one of the restrictions.

Reply:
I've specified that destStride and srcStride have to be non-zero.
I've added a bit to explain stride, I think it is quite clear.

I don't see a need to specify how the copy is done so I've tried to make it clear that it is implementation defined.